src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp

```diff
@@ -2626,19 +2627,110 @@ namespace OpenLoco::Ui::Windows::Construction::Construction
+        const uint8_t backupConstructionArrowDir = _constructionArrowDirection;
+
+        World::resetMapSelectionFlag(World::MapSelectionFlags::enableConstructionArrow);
+        if (_byte_522095 & (1 << 1))
+        {
+            World::setMapSelectionFlags(World::MapSelectionFlags::enableConstructionArrow);
+            _constructionArrowPos = pos;
+            _constructionArrowDirection = direction;
+        }
+
+        const auto* roadObj = ObjectManager::get<RoadObject>(roadType);
+        // Remove any none compatible road mods
+        for (auto mod = 0; mod < 2; ++mod)
+        {
+            if (selectedMods & (1 << mod))
+            {
+                const auto* roadExtraObj = ObjectManager::get<RoadExtraObject>(roadObj->mods[mod]);
+                if ((roadExtraObj->roadPieces & TrackData::getRoadMiscData(roadPieceId).compatibleFlags) != TrackData::getRoadMiscData(roadPieceId).compatibleFlags)
+                {
+                    selectedMods &= ~(1 << mod);
+                }
+            }
+        }
+
+        const auto& roadPieces = TrackData::getRoadPiece(roadPieceId);
+        const auto roadDirection = direction & 3;
+
+        World::TileElement backupTileElements[5] = {};
+
+        World::SurfaceElement previewSideSurfaceTileElement{ 255, 255, 0xF, true };
+        previewSideSurfaceTileElement.setLastFlag(true);
+
+        for (const auto& roadPiece : roadPieces)
+        {
+            const auto pieceOffset = World::Pos3{ Math::Vector::rotate(World::Pos2{ roadPiece.x, roadPiece.y }, roadDirection), roadPiece.z };
+            const auto quarterTile = roadPiece.subTileClearance.rotate(roadDirection);
+            const auto trackPos = pos + pieceOffset;
+            const auto baseZ = trackPos.z / kSmallZStep;
+            const auto clearZ = baseZ + (roadPiece.clearZ + 32) / kSmallZStep;
+
+            const auto centreTileCoords = World::toTileSpace(trackPos);
+            const auto eastTileCoords = centreTileCoords + World::toTileSpace(World::kOffsets[1]);
+            const auto westTileCoords = centreTileCoords - World::toTileSpace(World::kOffsets[1]);
+            const auto northTileCoords = centreTileCoords + World::toTileSpace(World::kOffsets[3]);
+            const auto southTileCoords = centreTileCoords - World::toTileSpace(World::kOffsets[3]);
+
+            // Copy map elements which will be replaced with temporary ones containing road
+            backupTileElements[0] = *World::TileManager::get(centreTileCoords)[0];
+            backupTileElements[1] = *World::TileManager::get(eastTileCoords)[0];
+            backupTileElements[2] = *World::TileManager::get(westTileCoords)[0];
+            backupTileElements[3] = *World::TileManager::get(northTileCoords)[0];
+            backupTileElements[4] = *World::TileManager::get(southTileCoords)[0];
+
+            // Set the temporary road element
+            World::RoadElement newRoadEl{};
+            newRoadEl.setType(World::ElementType::road);
```

Comment:
For track i made a constructor rather than used individual setters. What are peoples thoughts?

Reply:
I like that the constructor sets the type but I think as for individual properties it shouldn't be all passed by the constructor.

Reply:
if you need to set certain properties for an object to be considered 'valid', or they're otherwise mandatory values, then these should be part of the constructor. anything else can be setters

Reply:
What is a valid element is actually not that easily defined in CS land, I think that everything null initialized except the type will not really break the game or cause any issues which sort of makes most variables optional. I think passing everything over the constructor doesn't help readability, but I generally agree that an object after construction should be in a valid/useable state.

Reply:
I've gone for setting the type and the heights as the height is also used to mark a tile as invalid. (Clear height isn't but i thought best to keep them together)